docs/man/nut-ipmipsu.8
'\" t
.\"     Title: nut-ipmipsu
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 08/08/2025
.\"    Manual: NUT Manual
.\"    Source: Network UPS Tools 2.8.4
.\"  Language: English
.\"
.TH "NUT\-IPMIPSU" "8" "08/08/2025" "Network UPS Tools 2\&.8\&.4" "NUT Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
nut-ipmipsu \- Driver for IPMI Power Supply Units (PSU)
.SH "SYNOPSIS"
.sp
\fBnut\-ipmipsu\fR \-h
.sp
\fBnut\-ipmipsu\fR \-a \fIPSU_NAME\fR [\fIOPTIONS\fR]
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.sp
This driver is experimental, and still a work\-in\-progress\&. Feedback is encouraged\&.
.sp .5v
.RE
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.sp
This man page only documents the hardware\-specific features of the nut\-ipmipsu driver\&. For information about the core driver, see \fBnutupsdrv\fR(8)\&.
.sp .5v
.RE
.SH "SUPPORTED HARDWARE"
.sp
This driver should support a wide range of PSUs through local IPMI interface\&.
.sp
nut\-ipmipsu currently uses the GNU FreeIPMI project, for IPMI implementation\&.
.SH "EXTRA ARGUMENTS"
.sp
This driver doesn\(cqt support any optional settings\&.
.SH "INSTALLATION"
.sp
This driver may be not built by default\&. You can build it by installing the libfreeipmi\-dev dependencies and using configure \-\-with\-ipmi=yes\&.
.sp
You also need to give proper permissions on the local IPMI device file (/dev/ipmi0 for example) to allow the NUT user to access it\&.
.sp
An \fBudev\fR rules file (nut\-ipmipsu\&.rules) is provided and automatically installed on an \fBudev\fR enabled system\&. This file is generally installed in /etc/udev/rules\&.d/, or /lib/udev/rules\&.d/ on newer systems, to address the permission settings problem\&.
.sp
For more information, refer to scripts/udev/README\&.adoc in NUT sources\&.
.SH "INSTANT COMMANDS"
.sp
This driver doesn\(cqt support any instant commands\&.
.SH "IMPLEMENTATION"
.sp
The port value is used to identify the PSU\&. For instance, to target FRU \fI0x2\fR, use the following in \fBups\&.conf\fR:
.sp
.if n \{\
.RS 4
.\}
.nf
[pdu]
        driver = nut\-ipmipsu
        port = id2
.fi
.if n \{\
.RE
.\}
.sp
This driver will report various information related to a PSU, including:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
manufacturer, model, serial and part numbers,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
nominal voltage and frequency,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
actual current and voltage,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
status of the PSU:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIOL\fR
means that the PSU is present and providing power,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIOFF\fR
means that the PSU is present but not providing power (power cable removed),
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIstale\fR
(no data) means that the PSU is not present (i\&.e\&. physically removed)\&.
.RE
.RE
.sp
Here is an example output for a Dell r610 server:
.sp
.if n \{\
.RS 4
.\}
.nf
device\&.mfr: DELL
device\&.mfr\&.date: 01/05/11 \- 08:51:00
device\&.model: PWR SPLY,717W,RDNT
device\&.part: 0RN442A01
device\&.serial: CN179721130031
device\&.type: psu
driver\&.name: nut\-ipmipsu
driver\&.parameter\&.pollinterval: 2
driver\&.parameter\&.port: id2
driver\&.version: 2\&.6\&.1\-3139M
driver\&.version\&.data: IPMI PSU driver
driver\&.version\&.internal: 0\&.01
input\&.current: 0\&.20
input\&.frequency\&.high: 63
input\&.frequency\&.low: 47
input\&.voltage: 232\&.00
input\&.voltage\&.maximum: 264
input\&.voltage\&.minimum: 90
ups\&.id: 2
ups\&.realpower\&.nominal: 717
ups\&.status: OL
ups\&.voltage: 12
.fi
.if n \{\
.RE
.\}
.SH "KNOWN ISSUES"
.SS "Using host names for UPS NMC"
.sp
An UPS network management card may be assigned a fixed/static IP address or a dynamic one (e\&.g\&. by DHCP) in your network\&. Due to this, you may want or have to use a dynamic naming service to access the UPS\&. Note that this may become a problem specifically during large outages and shutdowns, when your DHCP/DNS server might already go down while the driver needs to resolve the name involved (especially during late\-shutdown hooks, when a new instance of the driver program might start just to tell the UPS to power off or to power\-cycle)\&.
.sp
It may be wise to ensure your OS name service client can cache the UPS name sufficiently long, or to use fixed IP addressing (and an entry in /etc/hosts for good measure, so you only have one spot to eventually re\-configure this)\&.
.SH "AUTHOR"
.sp
Arnaud Quette <arnaud\&.quette@free\&.fr>
.SH "SEE ALSO"
.SS "The core driver:"
.sp
\fBnutupsdrv\fR(8)
.SS "Internet resources:"
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
The NUT (Network UPS Tools) home page:
https://www\&.networkupstools\&.org/historic/v2\&.8\&.4/
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
GNU FreeIPMI home page:
http://www\&.gnu\&.org/software/freeipmi/
.RE


NEWS
Please see NEWS.adoc for actual contents


drivers/nut-ipmipsu.c
/*  nut-ipmipsu.c - Driver for IPMI Power Supply Units (PSU)
 *
 *  Copyright (C)
 *    2011 - 2012  Arnaud Quette <arnaud.quette@free.fr>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * TODO list:
 * - PSU sensor monitoring (how to find the right one?)
 * - dump all value at init, so that we can check for other interesting data
 */

#include "main.h"
#include "nut-ipmi.h"

#define DRIVER_NAME	"IPMI PSU driver"
#define DRIVER_VERSION	"0.35"

/* driver description structure */
upsdrv_info_t upsdrv_info = {
	DRIVER_NAME,
	DRIVER_VERSION,
	"Arnaud Quette <arnaud.quette@free.fr>\n",
	DRV_EXPERIMENTAL,
	{ NULL }
};

/* Note on device.status
 * OL: present and providing power
 * OFF: present but not providing power (power cable removed)
 * stale: not present (PSU removed)
 * 	=> should we prefer RB, MISSING, ABSENT, ???
 */

/* Abstract structure to allow different IPMI implementation
 * We currently use FreeIPMI, but OpenIPMI and others are serious
 * candidates! */
static IPMIDevice_t ipmi_dev;

/* Currently used to store FRU ID, but will probably evolve... */
static int ipmi_id = -1;

void upsdrv_initinfo(void)
{
	/* try to detect the PSU here - call fatal_with_errno(EXIT_FAILURE, ) if it fails */
	upsdebugx(1, "upsdrv_initinfo...");

	/* print what we detected during IPMI open */
	upsdebugx(1, "Detected a PSU: %s/%s",
		ipmi_dev.manufacturer ? ipmi_dev.manufacturer : "unknown",
		ipmi_dev.product ? ipmi_dev.product : "unknown");

	dstate_setinfo ("device.type", "psu");

	/* Publish information from the IPMI structure */
	if (ipmi_dev.manufacturer)
		dstate_setinfo("device.mfr", "%s", ipmi_dev.manufacturer);

	if (ipmi_dev.product)
		dstate_setinfo("device.model", "%s", ipmi_dev.product);

	if (ipmi_dev.serial)
		dstate_setinfo("device.serial", "%s", ipmi_dev.serial);

	if (ipmi_dev.part)
		dstate_setinfo("device.part", "%s", ipmi_dev.part);

	if (ipmi_dev.date)
		dstate_setinfo("device.mfr.date", "%s", ipmi_dev.date);

	/* FIXME: move to device.id */
	dstate_setinfo("ups.id", "%i", ipmi_id);
	/* FIXME: move to device.realpower.nominal */
	if (ipmi_dev.overall_capacity != -1)
		dstate_setinfo("ups.realpower.nominal", "%i", ipmi_dev.overall_capacity);

	/* FIXME: Did older FreeIPMI with "unsigned int" voltage ranges
	 * have a way to report invalid readings?
	 */
#ifdef HAVE_FREEIPMI_11X_12X
	if (ipmi_dev.input_minvoltage != -1)
#endif
		dstate_setinfo("input.voltage.minimum", "%i", ipmi_dev.input_minvoltage);

#ifdef HAVE_FREEIPMI_11X_12X
	if (ipmi_dev.input_maxvoltage != -1)
#endif
		dstate_setinfo("input.voltage.maximum", "%i", ipmi_dev.input_maxvoltage);

	if (ipmi_dev.input_minfreq != -1)
		dstate_setinfo("input.frequency.low", "%i", ipmi_dev.input_minfreq);

	if (ipmi_dev.input_maxfreq != -1)
		dstate_setinfo("input.frequency.high", "%i", ipmi_dev.input_maxfreq);

	/* FIXME: move to device.voltage */
	if (ipmi_dev.voltage != -1)
		dstate_setinfo("ups.voltage", "%i", ipmi_dev.voltage);

	if (nut_ipmi_monitoring_init() != 0)
		fatalx(EXIT_FAILURE, "Can't initialize IPMI monitoring");

	if (nut_ipmi_get_sensors_status(&ipmi_dev) != 0) {
		upsdebugx(1, "Error while updating sensors values");
		dstate_datastale();
	}
	else {
		dstate_dataok();
	}

	/* upsh.instcmd = instcmd; */
}

void upsdrv_updateinfo(void)
{
	upsdebugx(1, "upsdrv_updateinfo...");

	/* FIXME: implement sensors monitoring */

	if (nut_ipmi_get_sensors_status(&ipmi_dev) != 0) {
		upsdebugx(1, "Error while updating sensors values");
		dstate_datastale();
	}
	else {
		dstate_dataok();
	}

	/*
	 * poll_interval = 2;
	 */
}

void upsdrv_shutdown(void)
{
	/* Only implement "shutdown.default"; do not invoke
	 * general handling of other `sdcommands` here */

	/*
	 * WARNING:
	 * This driver will probably never support this properly:
	 * In order to be of any use, the driver should be called
	 * near the end of the system halt script (or a service
	 * management framework's equivalent, if any). By that
	 * time we, in all likelyhood, won't have basic network
	 * capabilities anymore, so we could never send this
	 * command to the UPS. This is not an error, but rather
	 * a limitation (on some platforms) of the interface/media
	 * used for these devices.
	 */

	/* replace with a proper shutdown function */
	upslogx(LOG_ERR, "shutdown not supported");
	if (handling_upsdrv_shutdown > 0)
		set_exit_flag(EF_EXIT_FAILURE);
}

/*
static int instcmd(const char *cmdname, const char *extra)
{
	/ * May be used in logging below, but not as a command argument * /
	NUT_UNUSED_VARIABLE(extra);
	upsdebug_INSTCMD_STARTING(cmdname, extra);

	if (!strcasecmp(cmdname, "test.battery.stop")) {
		upslog_INSTCMD_POWERSTATE_MAYBE(cmdname, extra);
		ser_send_buf(upsfd, ...);
		return STAT_INSTCMD_HANDLED;
	}

	upslog_INSTCMD_UNKNOWN(cmdname, extra);
	return STAT_INSTCMD_UNKNOWN;
}
*/

/*
static int setvar(const char *varname, const char *val)
{
	upsdebug_SET_STARTING(varname, val);

  	if (!strcasecmp(varname, "ups.test.interval")) {
		ser_send_buf(upsfd, ...);
		return STAT_SET_HANDLED;
	}

	upslog_SET_UNKNOWN(varname, val);
	return STAT_SET_UNKNOWN;
}
*/

void upsdrv_help(void)
{
}

/* list flags and values that you want to receive via -x */
void upsdrv_makevartable(void)
{
	/* FIXME: need more params. */
/*
	addvar(VAR_VALUE, "username", "Remote server username");
	addvar(VAR_VALUE, "password", "Remote server password");
	addvar(VAR_VALUE, "authtype",
			"Authentication type to use during lan session activation");
	addvar(VAR_VALUE, "type",
		"Type of the device to match ('psu' for \"Power Supply\")");

	addvar(VAR_VALUE, "serial", "Serial number to match a specific device");
	addvar(VAR_VALUE, "fruid", "FRU identifier to match a specific device");
	addvar(VAR_VALUE, "sensorid", "Sensor identifier to match a specific device");
*/
}

void upsdrv_initups(void)
{
	upsdebugx(1, "upsdrv_initups...");

	/* port can be expressed in various forms:
	 * - inband:
	 *   "id?" for device (FRU) ID 0x?
	 *   "psu?" for PSU number ?
	 * - out of band
	 *   "id?@host"
	 *   "host" => requires serial or ...
	 */
	if (!strncmp( device_path, "id", 2))
	{
		ipmi_id = atoi(device_path+2);
		upsdebugx(2, "Device ID 0x%i", ipmi_id);
	}
	/* else... <psuX> to select PSU number X */

	/* Clear the interface structure */
	ipmi_dev.ipmi_id = -1;
	ipmi_dev.manufacturer = NULL;
	ipmi_dev.product = NULL;
	ipmi_dev.serial = NULL;
	ipmi_dev.part = NULL;
	ipmi_dev.date = NULL;
	ipmi_dev.overall_capacity = -1;
	ipmi_dev.input_minvoltage = -1;
	ipmi_dev.input_maxvoltage = -1;
	ipmi_dev.input_minfreq = -1;
	ipmi_dev.input_maxfreq = -1;
	ipmi_dev.voltage = -1;
	ipmi_dev.sensors_count = 0;
	ipmi_dev.status = -1;
	ipmi_dev.input_voltage = -1;
	ipmi_dev.input_current = -1;
	ipmi_dev.temperature = -1;

	/* Open IPMI using the above */
	nut_ipmi_open(ipmi_id, &ipmi_dev);

	/* the upsh handlers can't be done here, as they get initialized
	 * shortly after upsdrv_initups returns to main.
	 */

	/* don't try to detect the UPS here */
}

void upsdrv_cleanup(void)
{
	upsdebugx(1, "upsdrv_cleanup...");
	nut_ipmi_close();
}
